### ISSUE TITLE
consider moving core sign & verify functions from cosign to sigstore

### ISSUE THREAD

**Issue by Dentrax** [sigstore/sigstore#81]
**TL;DR:** _I propose transferring the core crypto functions of cosign to project Sigstore._

**Motivation**

We aim to make sigstore more _commonize_ for cryptographic functions. Cosign calls the all the crypto-specific functions _(i.e., generating ECDSA, signing, validating, etc.)_ from the [keys.go](https://github.com/sigstore/cosign/blob/main/pkg/cosign/keys.go) file. If we get rid of those ECDSA-related functions, we would easily import sigstore project and use those functions again in cosign; moreover, we would import them on other _non-container-specific_ projects. For cosign, this work will likely legacy change for no discernible benefit. I think it should be better to handle these functions (like `GenerateKeyPair()`, `LoadECDSAPrivateKey()`, `LoadPublicKey()`, etc.) inside the Sigstore itself.

We could find neither an actively discussing issue nor a PR. It is a great opportunity to get used to Sigstore project.

**Implementation**

```go
func GeneratePrivateKey() (*ecdsa.PrivateKey, error)

func GenerateKeyPair(options *GenerateKeyPairOptions) (*KeyPair, error)

func LoadPublicKey(options *PublicKeyOptions) (PublicKey, error)

func LoadPrivateKey(options *PrivateKeyOptions) (signature.ECDSASignerVerifier, error)

func KeyToPem(pub crypto.PublicKey) ([]byte, error)

func CertToPem(c *x509.Certificate) []byte

func PemToECDSAKey(raw []byte) (*ecdsa.PublicKey, error)
```

Example Usages:

```go
package main

import (
	"github.com/sigstore/sigstore/pkg/signature"
)

func main() {
    s, _ := signature.GenerateKeyPair(&signature.GenerateKeyPairOptions{
        PassFunc := pf()
    })

    content := "foo"
    
    pub, _ := signature.LoadPublicKey(&signature.PublicKeyOptions{
        Path: "cosign.pub",
    })

    key, _ := signature.LoadPrivateKey(&signature.PrivateKeyOptions{
        Path: "cosign.key",
        Pass: []byte("foo"),
    })
    
    s1, s2, _ := key.Sign(context.TODO(), []byte(content))
    
    _ = pub.Verify(context.TODO(), []byte(content), s1)

    // ...
}
```

**Comment by bobcallaway** [comment]
great idea - now that #69 is merged, I will be pushing up a patch to cosign to take advantage of more of the code in `pkg/signature` instead of having duplicative function within cosign.

**Comment by lukehinds** [comment]
What a well written issue, kudos. I can see Bob just popped up before I started to type. The plan all along has been to put commonly shared functionality into sigstore/sigstore. I think the cosigners have just not got around to this yet. Coincidently tektoncd/chains also inherits from sigstore/sigstore too.

**Comment by Dentrax** [comment]
Thanks! Looks like this was a already planned task. So, we can analyze the _refactoring complexity_ before deep diving into this @developer-guy?

**Comment by developer-guy** [comment]
Yeah we definitely would want to get involved in it, we can be volunteers to do that with @Dentrax 🙋🏻‍♂️🤩

**Comment by lukehinds** [comment]
relates to: https://github.com/sigstore/sigstore/issues/89

**Comment by dekkagaijin** [comment]
If we want `sigstore/sigstore` to offer those kind of high-level libs (e.g. verifying both signatures and tlog entries), we'd probably want to move some of the more fundamental libs from s/s to their own repo in order to avoid circular dependencies.

I'd argue that container and `cosign`-implementation-specific logic should stay in `sigstore/cosign`, but we could easily generalize its signing and verification flows.

**Comment by lukehinds** [comment]
how are we with this now..is anyone active on the issue?

**Comment by lukehinds** [comment]
@Dentrax / @developer-guy do you plan to work on this?

**Comment by Dentrax** [comment]
Oh, sorry, we missed your last comment. 😇  If proposal accepted, and makes sense, then we can start to work on this with pleasure.

**Comment by lukehinds** [comment]
> Oh, sorry, we missed your last comment. innocent If proposal accepted, and makes sense, then we can start to work on this with pleasure.

I will assign to you and thanks!

It would make sense to tag this work for 1.0 as well.

**Comment by lukehinds** [comment]
@Dentrax / @developer-guy , we are planning on getting 1.0 out the door soon. I don't believe you need any changes in here, its all cosign side? Can you tag this issue when you make a PR to cosign?

**Comment by Dentrax** [comment]
> @Dentrax / @developer-guy , we are planning on getting 1.0 out the door soon. I don't believe you need any changes in here, its all cosign side? Can you tag this issue when you make a PR to cosign?

Yes we can tag this issue. But I think we need to move [keys.go](https://github.com/sigstore/cosign/blob/main/pkg/cosign/keys.go) from cosign to here (some funcs and interfaces), which means we may need to impl the core functions and make changes here. 🤔

**Comment by haydentherapper** [comment]
Closing as this will be a part of sigstore-go
